### PR TITLE
[framework] Add DUT connection by console port

### PR DESCRIPTION
**Pull request by bingwang-ms**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

#### What is the motivation for this PR?
As we all know, Ansible run commands via SSH connection from mgmt port. However, the connection is not stable or not usable for some scenes, like vrf test, reboot test and etc. Therefore, we need a more robust way to connect and run command on DUT. This commit add enables framework to connect DUT by console port.
#### How did you do it?

1. The feature is implemented based on netmiko.
 Cueerntly, two types of console connection in star lab are supported, that is console on Digi (**digi_console**) and console servers using telnet. Some interfaces are needed to be overwriten if more console types are to support.
2.  The console type, address and port of console hub are defined in inventory file, 
```
str-7260cx3-acs-1:
      ansible_host: 10.64.247.224
      hwsku: Arista-7260CX3-D108C8
      pdu_host: pdu-106
      console_type: telnet_console
      console_host: 10.3.146.16
      console_port: 2015
```  
and console username and password are managed by Ansible secret vars.
3. Add a new fixture duthost_console for establishing connection to DUT by console port. 
4. All public interfaces are inherited from netmiko. But only **send_command** is tested for now.
[https://ktbyers.github.io/netmiko/docs/netmiko/index.html#netmiko.BaseConnection.send_command](url)
#### How did you verify/test it?
I write a demo script to verify it
```
import pytest
import logging

pytestmark = [
    pytest.mark.topology('t0'),
]

def test_telnet(duthost_console):
    out = duthost_console.send_command("show version")
    logging.info(out)
```

#### Any platform specific information?
No.
#### Supported testbed topology if it's a new test case?
N/A.
### Documentation 

Please refer to Netmiko doc 
[https://ktbyers.github.io/netmiko/docs/netmiko/index.html](url)

### CODE REVIEW

File: tests/common/connections/ssh_console_conn.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            raise ValueError("Either console_server__username or console_server_password is not set")
+
+        # Console via SSH connection need two groups of user/passwd
+        self.sonic_username = kwargs['username']
+        self.sonic_password = kwargs['password']
+        console_username = kwargs['console_server_username'] + r':p=' + str(kwargs['port'])
+        # Restore the default port for SSH
+        del kwargs['port']
+        kwargs['username'] = console_username
+        kwargs['password'] = kwargs['console_server_password']
```
Comment (vaibhavhd) [inline tests/common/connections/ssh_console_conn.py:19]: How do we handle console connections that require Active Directory password. This is the case for most of the new DUTs (for eg. DX10s) that are onboarded and some Fanout switches. I think we will not want to have AD creds in secrets vars file.

File: tests/common/connections/ssh_console_conn.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                      pri_prompt_terminator=r".*# ",
+                      alt_prompt_terminator=r".*\$ ",
+                      username_pattern=r"(?:user:|username|login|user name)",
+                      pwd_pattern=r"assword",
+                      delay_factor=1,
+                      max_loops=20,
+                      ):
+        """
+        Perform a further login
+        :return: None
```
Comment (vaibhavhd) [inline tests/common/connections/ssh_console_conn.py:41]: This method seems to be returning `return_msg`

File: tests/common/connections/ssh_console_conn.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                    output = self.read_channel()
+                    return_msg += output
+
+                # Search for password pattern / send password
+                if re.search(pwd_pattern, output, flags=re.I):
+                    self.write_channel(self.sonic_password + self.TELNET_RETURN)
+                    time.sleep(0.5 * delay_factor)
+                    output = self.read_channel()
+                    return_msg += output
+                    if re.search(
```
Comment (vaibhavhd) [inline tests/common/connections/ssh_console_conn.py:68]: This condition is repeating in the line 74. Is there a reason to check twice?

File: tests/common/connections/telnet_console_conn.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+    def session_preparation(self):
+        super(TelnetConsoleConn, self).session_preparation()
+
+    def telnet_login(
+        self,
+        pri_prompt_terminator=r".*# ",
+        alt_prompt_terminator=r".*\$ ",
+        username_pattern=r"(?:user:|username|login|user name)",
+        pwd_pattern=r"assword",
```
Comment (vaibhavhd) [inline tests/common/connections/telnet_console_conn.py:16]: Typo in `assword`?

File: tests/common/connections/ssh_console_conn.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        output = ""
+        return_msg = ""
+        i = 1
+        while i <= max_loops:
+            try:
+                output = self.read_channel()
+                return_msg += output
+
+                # Search for username pattern / send username
+                if re.search(username_pattern, output, flags=re.I):
```
Comment (vaibhavhd) [inline tests/common/connections/ssh_console_conn.py:56]: I think we should gracefully handle the case when the line is busy. In that case the username_pattern will not be found.

File: tests/common/connections/telnet_console_conn.py (hunk 4 above)
Comment (yxieca) [reply tests/common/connections/telnet_console_conn.py:16]: I think this is work work with leading 'P' and 'p'

File: tests/common/connections/ssh_console_conn.py (hunk 2 above)
Comment (bingwang-ms) [reply tests/common/connections/ssh_console_conn.py:41]: Thanks. I missed it. I will post an update to fix it.

File: tests/common/connections/ssh_console_conn.py (hunk 5 above)
Comment (bingwang-ms) [reply tests/common/connections/ssh_console_conn.py:56]: Yes. I agree with you. But there is nothing we can do except waiting for session timeout. There is no output at that case.

File: tests/common/connections/ssh_console_conn.py (hunk 3 above)
Comment (bingwang-ms) [reply tests/common/connections/ssh_console_conn.py:68]: Good catch. The check in ln68-71 is not necessary. But the check in ln74-77 must be kept to handle the case when password is not required to login.

File: tests/common/connections/ssh_console_conn.py (hunk 5 above)
Comment (vaibhavhd) [reply tests/common/connections/ssh_console_conn.py:56]: Oh! In that case, can we do it in two steps? 1 - Login to the console switch 2- Check if the line is busy?

File: tests/common/connections/ssh_console_conn.py (hunk 1 above)
Comment (bingwang-ms) [reply tests/common/connections/ssh_console_conn.py:19]: Great question. The auth method used in Lab highly depends on Microsoft AD. Perhaps we could try to add a public account for test.

File: tests/common/connections/ssh_console_conn.py (hunk 5 above)
Comment (bingwang-ms) [reply tests/common/connections/ssh_console_conn.py:56]: I'm not sure if it's possible. For some console server (like mad console), there will be nothing output if the port is busy.  We have no idea whether it's cause by latency or busy line.

File: tests/conftest.py [hunk 6]
```diff
@@ -427,3 +430,20 @@ def tag_test_report(request, pytestconfig, testbed, duthost, record_testsuite_pr
...
+    dut_hostname = testbed["duts"][0]
+
+    vars = localhost.host.options['inventory_manager'].get_host(dut_hostname).vars
+    host = ConsoleHost(console_type=vars['console_type'],
+                       console_host=vars['console_host'],
+                       console_port=vars['console_port'],
+                       sonic_username=creds['sonicadmin_user'],
+                       sonic_password=creds['sonicadmin_password'],
+                       console_server_username=creds['console_server_user'],
+                       console_server_password=creds['console_server_password'])
```
[the remark this answers is not in the document]
Comment (bingwang-ms) [reply tests/conftest.py:446]: Good catch. I need to come up with a better method to manage account info for various console server.

**Review by lguohan — APPROVED** [review]

**Comment by lguohan** [comment]
most of the part looks good to me. maybe we can get it merged first and then iterate?

**Comment by bingwang-ms** [comment]
> most of the part looks good to me. maybe we can get it merged first and then iterate?

Thanks. Now I'm waiting for this PR https://github.com/Azure/sonic-mgmt/pull/2257 (Pdu and Console connection info in connection graph), which puts console meta data in connection graph. Currently, I put them in inventory file. But it shouldn't be too much work.

**Comment by bingwang-ms** [comment]
@yxieca @vaibhavhd  @wangxin Sorry for the late update. Please help to review again.
Major updates:
1. Support console login via ```menu ports```
2. Support multi password for console server and SONiC

Please note that the current code reads console information from inventory file and reads creds from secret file. The next step is to put console information into connect_graph_facts. I will work with  @sujinmkang to complete PR https://github.com/Azure/sonic-mgmt/pull/2257

**Review by wangxin — APPROVED** [review]

**Pull request merged**